## Jacknife

El método Jacknife fue propuesto para estimar errores estándar y sesgos.

Dada una variable muestral $X_j=(x_{j}, x_{2j},..., x_{nj})^t\hspace{0.2cm}$  y un estidímador $\hspace{0.2cm}\widehat{\theta}= f(X_j)$ del parametro poblacional $\theta$

El sesgo del estimador $\widehat{\theta}$ se define como:

$$Sesgo(\widehat{\theta}) = E \left[ \hspace{0.05cm} \widehat{\theta} \hspace{0.05cm} \right]- \theta$$

La varianza del estimador $\widehat{\theta}$ se define como:

$$Var(\widehat{\theta}) = E \left[ \hspace{0.05cm} \left( \widehat{\theta} - E [ \hspace{0.05cm} \widehat{\theta} \hspace{0.05cm}  ] \right)^2 \hspace{0.05cm} \right]$$

El error estandar (desviación típica) del estimador $\widehat{\theta}$ se define como:

$$s.e.(\widehat{\theta}) = \sqrt{Var(\widehat{\theta})}$$

Se define la muestra $r$-esima de $X_j$ como:

$$X_{(r)j} = (x_{1j}, x_{2j},..,x_{r-1,j},x_{r+1,j} ,.., x_{nj}) \hspace{0.4cm}$$  

para  $\hspace{0.2cm} r=1,...,n$

Se define la replica $r$-esima del estimador $\widehat{\theta}$ como:

$$\hat{\theta}_{(r)} = f(X_{(r)j})$$

El **estimador Jacknife** del **sesgo** del estimador $\widehat{\theta}$ se define como:

$$\widehat{Sesgo}(\widehat{\theta})_{Jack} =  (n-1) \cdot \left(  \widehat{\theta}_{(\cdot)} - \widehat{\theta} \right)$$

Donde: 

$$\widehat{\theta}_{(\cdot)} = \dfrac{1}{n} \cdot \sum_{r=1}^{n} \widehat{\theta}_{(r)}$$

El **estimador Jacknife** de la **varianza** del estimador $\widehat{\theta}$ se define como:

$$\widehat{ Var}(\widehat{\theta})_{Jack} =  \dfrac{n-1}{n} \cdot \sum_{r=1}^{n} \left( \widehat{\theta}_{(r)} - \dfrac{1}{n} \sum_{r=1}^{n} \widehat{\theta}_{(r)}  \right)^2$$

El **estimador Jacknife** del **error estandar** del estimador $\widehat{\theta}$ se define como:

$$\widehat{ s.e.}(\widehat{\theta})_{Jack} = \sqrt{ \dfrac{n-1}{n} \cdot \sum_{r=1}^{n} \left( \widehat{\theta}_{(r)} - \dfrac{1}{n} \sum_{r=1}^{n} \widehat{\theta}_{(r)}  \right)^2}$$

**Observación:**

El Jacknife funciona bien cuando el estimador es suave (smooth). 

Un estimador es suave cuando ante pequeños cambios en la muestra de datos genera pequeños cambios en el estimador.

Ejemplo de estimador suave es el estimador plug-in de la media poblacional, es decir la media muestral.

Ejemplo de estimador no suave es el estimador plug-in de la mediana poblacional, es decir la mediana muestral

**Estimación Jacknife con corrección de sesgo**

\begin{gather*}
\widehat{\theta}_{Jack} \hspace{0.2cm}=\hspace{0.2cm} \widehat{\theta} \hspace{0.1cm}-\hspace{0.1cm} \widehat{Sesgo}(\widehat{\theta})_{Jack} \\ \\ \hspace{0.5cm}=\hspace{0.2cm} \widehat{\theta} -  (n-1) \cdot \left(  \widehat{\theta}_{(\cdot)} - \widehat{\theta} \right) \\  \\ \hspace{0.4cm}=\hspace{0.2cm}   n\cdot \widehat{\theta} - (n-1)\cdot \widehat{\theta}_{(\cdot)}
\end{gather*}

### Aplicación del Jacknife en `Python`

In [2]:
import numpy as np

In [3]:
np.random.seed(123)

Xj = np.random.normal(loc=10, scale=15, size=50)

In [4]:
Xj

array([ -6.28445905,  24.9601817 ,  14.24467747, -12.59442071,
         1.32099622,  34.77154806, -26.40018865,   3.56631057,
        28.98904388,  -3.00110603,  -0.18329227,   8.57936547,
        32.37084439,   0.41647005,   3.34027061,   3.48473087,
        43.08895124,  42.80179133,  25.06080847,  15.79279599,
        21.06052864,  32.36098042,  -4.03750803,  27.63743567,
        -8.80821002,   0.43372746,  23.60657794, -11.4302105 ,
         7.8989692 ,  -2.92632344,   6.16570944, -31.97883658,
       -16.57299657,  -0.49815852,  23.91193648,   7.39546476,
        10.04268874,  20.32334067,  -3.19304515,  14.25440986,
        -2.08049777, -15.91504241,   4.13650309,  18.60708794,
        15.07883576,   9.82254258,  45.88547899,  16.1936824 ,
        24.68104009,  43.57215008])

In [5]:
np.delete(Xj, 1)

array([ -6.28445905,  14.24467747, -12.59442071,   1.32099622,
        34.77154806, -26.40018865,   3.56631057,  28.98904388,
        -3.00110603,  -0.18329227,   8.57936547,  32.37084439,
         0.41647005,   3.34027061,   3.48473087,  43.08895124,
        42.80179133,  25.06080847,  15.79279599,  21.06052864,
        32.36098042,  -4.03750803,  27.63743567,  -8.80821002,
         0.43372746,  23.60657794, -11.4302105 ,   7.8989692 ,
        -2.92632344,   6.16570944, -31.97883658, -16.57299657,
        -0.49815852,  23.91193648,   7.39546476,  10.04268874,
        20.32334067,  -3.19304515,  14.25440986,  -2.08049777,
       -15.91504241,   4.13650309,  18.60708794,  15.07883576,
         9.82254258,  45.88547899,  16.1936824 ,  24.68104009,
        43.57215008])

In [76]:
def Jacknife(Xj , estimador, q=0.75):

#######################################################

    if estimador == 'median':

        def Jacknife_sample(Xj , r):

            Xj_sample_r = np.delete(Xj, r)

            return(Xj_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(Xj)):

            replicas_estimador.append( np.median( Jacknife_sample(Xj , r) ) )

    
        n = len(Xj)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.median(Xj) )

        estimacion_sesgo_corregido = n*np.median(Xj) - (n-1)*np.mean( replicas_estimador )

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )

 #######################################################   

    if estimador == 'mean':

        def Jacknife_sample(Xj , r):

            Xj_sample_r = np.delete(Xj, r)

            return(Xj_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(Xj)):

            replicas_estimador.append( np.mean( Jacknife_sample(Xj , r) ) )

    
        n = len(Xj)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.mean(Xj) )

        estimacion_sesgo_corregido = n*np.mean(Xj) - (n-1)*np.mean( replicas_estimador )

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################

    if estimador == 'std':

        def Jacknife_sample(Xj , r):

            Xj_sample_r = np.delete(Xj, r)

            return(Xj_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(Xj)):

            replicas_estimador.append( np.std( Jacknife_sample(Xj , r) ) )

    
        n = len(Xj)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.std(Xj) )

        estimacion_sesgo_corregido = n*np.std(Xj) - (n-1)*np.mean( replicas_estimador )

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################

    if estimador == 'quantile':

        def Jacknife_sample(Xj , r):

            Xj_sample_r = np.delete(Xj, r)

            return(Xj_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(Xj)):

            replicas_estimador.append( np.quantile( Jacknife_sample(Xj , r) , q) )

    
        n = len(Xj)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - np.quantile(Xj , q) )

        estimacion_sesgo_corregido = n*np.quantile(Xj , q) - (n-1)*np.mean( replicas_estimador )

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )

#######################################################

    if estimador == 'kurtosis':

        import scipy

        from scipy.stats import  kurtosis

        def Jacknife_sample(Xj , r):

            Xj_sample_r = np.delete(Xj, r)

            return(Xj_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(Xj)):

            replicas_estimador.append( kurtosis(Xj) )

    
        n = len(Xj)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - kurtosis(Xj) )

        estimacion_sesgo_corregido = n*kurtosis(Xj) - (n-1)*np.mean( replicas_estimador )

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
     
#######################################################

    if estimador == 'asimetria':

        import scipy

        from scipy.stats import  skew

        def Jacknife_sample(Xj , r):

            Xj_sample_r = np.delete(Xj, r)

            return(Xj_sample_r)

        
        replicas_estimador = []

        for r in range(0, len(Xj)):

            replicas_estimador.append( skew(Xj) )

    
        n = len(Xj)

        sesgo = (n-1) * ( np.mean( replicas_estimador ) - skew(Xj) )

        estimacion_sesgo_corregido = n*skew(Xj) - (n-1)*np.mean( replicas_estimador )

        standard_error = np.sqrt( ((n-1)/n) * sum( (replicas_estimador - np.mean( replicas_estimador ))**2 ) )


#######################################################
    
    return(sesgo, estimacion_sesgo_corregido, standard_error)


In [71]:
import scipy

from scipy.stats import  kurtosis

In [73]:
kurtosis(Xj)

-0.37420768292897266

In [7]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'median', q=0.75)

In [8]:
sesgo

8.704148513061227e-14

In [9]:
estimacion_sesgo_corregido

8.239167331558235

In [10]:
standard_error

2.381386940718188

In [53]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'mean', q=0.75)

In [54]:
sesgo

-8.704148513061227e-14

In [55]:
estimacion_sesgo_corregido

10.199071616256902

In [56]:
standard_error

2.5491917443460235

In [63]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'std')

In [64]:
np.std(Xj)

17.844342210422187

In [65]:
sesgo

-0.26116994703598806

In [66]:
estimacion_sesgo_corregido

18.10551215745818

In [67]:
standard_error

1.6795955569730596

In [15]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'quantile', q=0.75)

In [19]:
np.quantile(Xj, 0.75)

23.835596841535178

In [16]:
sesgo

-0.14962568429344003

In [17]:
estimacion_sesgo_corregido

23.985222525828704

In [18]:
standard_error

0.9375849844484524

In [20]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'quantile', q=0.25)

In [24]:
np.quantile(Xj, 0.25)

-1.6849129579803668

In [21]:
sesgo

0.7753462334875716

In [22]:
estimacion_sesgo_corregido

-2.460259191467941

In [23]:
standard_error

4.858477270796004

In [25]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'quantile', q=0.10)

In [29]:
np.quantile(Xj, 0.10)

-11.546631523927719

In [26]:
sesgo

7.770186286477228

In [27]:
estimacion_sesgo_corregido

-19.31681781040504

In [28]:
standard_error

5.217098648611869

In [30]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'quantile', q=0.90)

In [34]:
np.quantile(Xj, 0.90)

32.61091475832362

In [31]:
sesgo

9.364358252194954

In [32]:
estimacion_sesgo_corregido

23.246556506128627

In [33]:
standard_error

1.0245278739126102

In [77]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'kurtosis')

In [78]:
sesgo

0.0

In [79]:
estimacion_sesgo_corregido

-0.3742076829289722

In [80]:
standard_error

0.0

In [81]:
sesgo, estimacion_sesgo_corregido, standard_error = Jacknife(Xj , 'asimetria')

In [82]:
sesgo

-1.700029006457271e-16

In [83]:
estimacion_sesgo_corregido

0.025587358812510397

In [84]:
standard_error

2.42861286636753e-17

## Contrastes de hipotesis por permutaciones

El test de permutaciones es un test de significancia estadística para el estudio de diferencias entre grupos. Fue desarrollado por Ronald Fisher y E.J.G. Pitman en 1930. La distribución del estadístico estudiado (media, mediana...) se obtiene calculando el valor de dicho estadístico para todas las posibles reorganizaciones de las observaciones en los distintos grupos. Dado que implica calcular todas las posibles situaciones, se trata de un test exacto.



La filosofia de los test de permutaciones es la siguiente:



Se tiene un conjunto de individuos, los cuales se distribuyen aleatoriamente en dos grupos, $T$ y $C$, de tamaños  $n_t$  y  $n_c$



Se aplica un experimento, que generalmente consiste en aplicar un tratamiento sobre un grupo (grupo tratamiento $T$) y no aplicarlo sobre el otro grupo (grupo control $C$).



Se mide una variable de interés $X$ en cada uno de los grupos:

$$ X^t = (x^t_{1},...,x^t_{n_t }) $$
$$ X^c = (x^c_{1},...,x^c_{n_c }) $$

Donde: 

$x^t_{i}$ es el valor de $X$ para el $i$-esimo individuo del grupo $T$

$x^c_{i}$ es el valor de $X$ para el $i$-esimo individuo del grupo $C$




Se calcula un estadistico $\widehat{\theta}$ de $X$ en los grupos $\Rightarrow \widehat{\theta}_t = f(X^t), \widehat{\theta}_c = f(X^c)$

Estadisticos habituales son la media y la mediana.



Se desea  determinar si existen diferencias significativas entre el valor del estadistico para cada grupo. 

Para ello se considera que si la diferencia observada en valor absoluto ($\mid D_{obs} \mid = \mid \widehat{\theta}_t - \widehat{\theta}_c \mid$ ) es 
mayor que la diferencia esperada por la asignación aleatoria de los grupos, en valor absoluto.
 

En términos estadísticos, la hipótesis nula ( $H_0$ ) es que ambas muestras pertenecen a la misma distribución y que, las diferencias observadas, son debidas únicamente a variaciones causadas por el reparto aleatorio en grupos. 



El test de permutación permite identificar si hay evidencias en contra de esta hipótesis. Para ello los pasos a seguir son:

1) Calcular la diferencia observada del estadistico para los dos grupos  

 $$\mid D_{obs} \mid = \mid \widehat{\theta}_t - \widehat{\theta}_c \mid$$



2) Todas las observaciones se unen sin tener en cuenta el grupo al que pertenecen. $(X^t , X^c)$



3) Se generan $B$ pares de muestras  sin reemplazamiento  de $(X^t , X^c)$ , de tamaño $n_t$ y $n_c$ respectivamente :

$$(\mathcal{X}^t_1 , \mathcal{X}^c_1), (\mathcal{X}^t_2 , \mathcal{X}^c_2),\dots, (\mathcal{X}^t_B , \mathcal{X}^c_B)$$


   $\mathcal{X}^g_i$ es la muestra aleatoria $i$-esima de tamaño n_g extraida de  $(X^t , X^c)$ , para $g=t,c$ , $i=1,...,B$
   
  



4) Para cada combinación, calcular la diferencia entre medias  (difcalculada) . El conjunto de valores obtenidos forman la distribución exacta de las posibles diferencias siendo cierta la hipótesis nula de que los dos grupos proceden de la misma población. A esta distribución, se le conoce como permutation distribution of the mean difference.

5) Calcular el p-value de dos colas como la proporción de combinaciones en las que, el valor absoluto de la diferencia calculada, es mayor o igual al valor absoluto de la diferencia observada. La idea es que si esta proporcion no es suficientemente pequeña, habrá evidencias de que las diferencias observadas entre los grupos son fruto del azar involucrado al generar los grupos y no del tratamiento.  Si en cambio esta proporcion es suficientemente pequeña, habra evidencias de que las doferencias observadas son causadas por el tratamiento, y no por la configuracion aleatoria de los grupos. Pero hay que delimitar un punto a partir del cual considerar esta proporcion como suficientmiente pequeña o no.

6) Si el p-value calculado en 5) es mayor que cierto limite $\alpha$ , no se podra rechazar H_0 en favor de H_1 puesto que se entiende que no hay suficiente evidencia. En caso de p-value menor que \alpha se acepta H_1 dada la evidencia dispobible. Normalmente se considera \alpha = 0.05 , 0.1 0 o 0.01 , en funcion del nivel de evidencia exigido. A mayor alpha, menos evidencia se exige para aceptar H_1 , cuando menor alpha mayor evidencia exigida para aceptar H_1

En este ejemplo se emplea como estadístico la media pero podría ser cualquier otro (mediana, varianza...)

Para que el test sea exacto deberia tomarse $B= \dfrac{(n_t+n_c)!}{n_t! \hspace{0.1cm} n_c!}$ , que es el numero de diferentes combinaciones  para distribuir $n_t + n_c$  observaciones en dos grupos de tamaños $n_t$ y $n_c$ 

Y ademas las B pares de muestras extraidas deberian de ser distintos (sin que importe el orden de los elementos, es decir dos muestras con los mismos elementos pero en diferente orden son consideradas iguales).

El problema de este enfoque exacto es que para tamaños moderados de $n_t + n_c$ se llegan a valores inasumibles de $B= \dfrac{(n_t+n_c)!}{n_t! \hspace{0.1cm} n_c!}$

Por ejemplo:

In [1]:
from math import factorial

In [5]:
factorial(30) / (factorial(20)*factorial(10))

30045015.0

In [7]:
factorial(30) / (factorial(15)*factorial(15))

155117520.0

La única condición necesaria para un test de permutaciones se conoce como exchangeability, según la cual, todas las posibles permutaciones tienen la misma probabilidad de ocurrir siendo cierta la hipótesis nula.

Los test de permutación son test de significancia y, por lo tanto, se emplean para calcular p-values. Es importante tener en cuenta que, las conclusiones extraídas, solo son aplicables a diseños de tipo experimental. Estudios en los que, tras haber elegido los sujetos, se realiza una asignación aleatoria a los diferentes grupos.

Nota: si bien se utiliza el término permutación, se trata realmente de combinaciones, ya que el orden de las observaciones dentro de cada grupo no importa.

Simulación de Monte Carlo

En los test de permutación, el p-value obtenido es exacto ya que se calculan todas las posibles combinaciones de las observaciones. Esto resulta complicado o imposible cuando el tamaño muestral es mediano o grande. Por ejemplo, existen más de 155 millones combinaciones distintas para distribuir 30 observaciones en dos grupos de 15 cada uno.



(m+n)!m!n!=(15+15)!15!15!=155117520
 


In [ ]:
# Número total de combinaciones
# ==============================================================================
from math import factorial
print(f"Total combinaciones = {factorial(15 + 15) / (factorial(15)*factorial(15))}")



Total combinaciones = 155117520.0
La simulación de Monte Carlo o randomization test consiste en emplear únicamente una muestra aleatoria de todas las posibles combinaciones, evitando así tener que calcularlas todas. Al no calcularse todas las posibles combinaciones, no es un test exacto sino aproximado.

El método de Monte Carlo no es insesgado. Una de las correcciones más empleadas es la propuesta por Davison and Hinkley):

pvalue=r+1k+1
 
siendo r el número de combinaciones igual o más extremas que el estadístico observado y k el número de combinaciones empleadas en la simulación.

A pesar de que la corrección es mínima cuando el número de simulaciones es alto, presenta la ventaja de que, si el valor observado es mayor que cualquiera de los calculados, el p-value es muy bajo pero no cero. Por facilidades de cálculo, se suele emplear un número de permutaciones terminado en 9 (4999, 9999).

¿Cuántas permutaciones (combinaciones) son necesarias?

No hay un número predefinido de combinaciones que asegure buenos resultados. Una forma de guiar la elección es tener en cuenta el valor mínimo capaz de ser detectado. Por ejemplo, con 1000 permutaciones, el menor valor de p-values posible es  pvalue=1+11000+1=0.001998 .

A pesar de ello, la estabilidad y el error de la estimación varían mucho en función de los datos disponibles. Suele encontrarse como recomendación empírica emplear al menos 2000 iteraciones.

¿Cómo interpretar el p-value obtenido?

El p-value es la probabilidad de obtener un valor del estadístico igual o más extremo que el valor observado, si la hipótesis nula es cierta  P(datos|H0) . Es importante no confundir esto con la probabilidad de que la hipótesis nula sea cierta  P(H0|datos) . Si el valor es muy pequeño, se puede interpretar como una evidencia en contra de la hipótesis nula, pero no como una confirmación de las hipótesis alternativas.

## Bibliografia

apuntes Juanmi

https://www.cienciadedatos.net/documentos/pystats03-test-permutacion-python.html